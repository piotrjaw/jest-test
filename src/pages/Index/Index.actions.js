import * as TYPES from './Index.types';
import { del, get, put, post } from '../../helpers/request';

const createTodoPending = () => ({
  type: TYPES.CREATE_TODO_PENDING
});

const createTodoSuccess = () => ({
  type: TYPES.CREATE_TODO_SUCCESS
});

const deleteTodoPending = () => ({
  type: TYPES.DELETE_TODO_PENDING
});

const deleteTodoSuccess = () => ({
  type: TYPES.DELETE_TODO_SUCCESS
});

const getTodosPending = () => ({
  type: TYPES.GET_TODOS_PENDING
});

export const getTodosSuccess = (data) => ({
  type: TYPES.GET_TODOS_SUCCESS,
  data
});

const updateTodoPending = () => ({
  type: TYPES.UPDATE_TODO_PENDING
});

const updateTodoSuccess = () => ({
  type: TYPES.UPDATE_TODO_SUCCESS
});

export const createTodo = (createObj) => (dispatch) => {
  if (!createObj.text) return Promise.reject();
  dispatch(createTodoPending());
  return post('/todos/', createObj)
    .then((res) => dispatch(createTodoSuccess()));
}

export const deleteTodo = (deleteObj) => (dispatch) => {
  dispatch(deleteTodoPending());
  return del('/todos/', deleteObj)
    .then((res) => dispatch(deleteTodoSuccess()));
};

export const getTodos = () => (dispatch) => {
  dispatch(getTodosPending());
  return get('/todos/')
    .then(({ data }) => dispatch(getTodosSuccess(data)));
};

export const updateTodo = (updateObj) => (dispatch) => {
  dispatch(updateTodoPending());
  return put('/todos/', updateObj)
    .then((res) => dispatch(updateTodoSuccess()));
};
