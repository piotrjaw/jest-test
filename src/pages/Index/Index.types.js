export const CREATE_TODO_PENDING = 'Index/CREATE_TODO_PENDING';
export const CREATE_TODO_SUCCESS = 'Index/CREATE_TODO_SUCCESS';
export const DELETE_TODO_PENDING = 'Index/DELETE_TODO_PENDING';
export const DELETE_TODO_SUCCESS = 'Index/DELETE_TODO_SUCCESS';
export const GET_TODOS_PENDING = 'Index/GET_TODOS_PENDING';
export const GET_TODOS_SUCCESS = 'Index/GET_TODOS_SUCCESS';
export const UPDATE_TODO_PENDING = 'Index/UPDATE_TODO_PENDING';
export const UPDATE_TODO_SUCCESS = 'Index/UPDATE_TODO_SUCCESS';
