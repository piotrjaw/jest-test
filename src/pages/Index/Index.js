import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as IndexActions from './Index.actions';

class IndexComponent extends PureComponent {
  render() {
    return (
      <div />
    );
  }
}

const mapStateToProps = ({ index }) => ({
  todos: index.todos
});

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators(IndexActions, dispatch)
});

const Index = connect(
  mapStateToProps,
  mapDispatchToProps
)(IndexComponent);

export default Index;
