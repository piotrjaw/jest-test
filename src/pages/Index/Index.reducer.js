import * as TYPES from './Index.types';

const initialState = { todos: [] };

export default (state = initialState, action) => {
  switch (action.type) {
    default:
      return state;
  }
};
