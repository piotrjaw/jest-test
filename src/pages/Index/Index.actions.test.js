import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';
import nock from 'nock';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as actions from './Index.actions';
import * as TYPES from './Index.types';

const middleWares = [thunk];
const mockStore = configureMockStore(middleWares);

const host = 'http://localhost:5000';

axios.defaults.adapter = httpAdapter;
axios.defaults.host = host;

let testActions;
let store;

beforeEach(() => {
  store = mockStore({ todos: [] });
  expect.assertions(2);
});

afterEach(() => {
  expect(testActions).toMatchSnapshot();
  nock.cleanAll();
});

describe('Index actions', () => {
  it('should create an action to create a todo', () => {
    const createObj = {
      text: 'test'
    };

    nock(host)
      .post('/todos/', createObj)
      .reply(200);

    const expectedActions = [
      { type: TYPES.CREATE_TODO_PENDING },
      { type: TYPES.CREATE_TODO_SUCCESS }
    ];

    testActions = store.getActions();

    return store
      .dispatch(actions.createTodo(createObj))
      .then(() => expect(testActions).toEqual(expectedActions));
  });

  it('should reject on lack of text', () => {
    const createObj = {
      text: undefined
    };

    testActions = null;

    nock(host)
      .post('/todos/', createObj)
      .reply(200);

    return expect(store.dispatch(actions.createTodo(createObj))).rejects.toBeUndefined();
  });

  it('should create an action to get all the todos', () => {
    nock(host)
      .get('/todos/')
      .reply(200, [
        {
          id: '1',
          text: '1',
          checked: false
        },
        {
          id: '2',
          text: '2',
          checked: true
        }
      ]);
    const expectedActions = [
      { type: TYPES.GET_TODOS_PENDING },
      { type: TYPES.GET_TODOS_SUCCESS, data: [
        {
          id: '1',
          text: '1',
          checked: false
        },
        {
          id: '2',
          text: '2',
          checked: true
        }
      ] }
    ];

    testActions = store.getActions();

    return store
      .dispatch(actions.getTodos())
      .then(() => expect(testActions).toEqual(expectedActions));
  });

  it('should create an action to udpate a todo', () => {
    const updateObj = {
      id: 1,
      text: 'test',
      checked: true
    };

    nock(host)
      .put('/todos/', updateObj)
      .reply(200);

    const expectedActions = [
      { type: TYPES.UPDATE_TODO_PENDING },
      { type: TYPES.UPDATE_TODO_SUCCESS }
    ];

    testActions = store.getActions();

    return store
      .dispatch(actions.updateTodo(updateObj))
      .then(() => expect(testActions).toEqual(expectedActions));
  });

  it('should create an action to delete a todo', () => {
    const deleteObj = { _id: 1 };
    nock(host)
      .delete('/todos/')
      .query(deleteObj)
      .reply(200);

    const expectedActions = [
      { type: TYPES.DELETE_TODO_PENDING },
      { type: TYPES.DELETE_TODO_SUCCESS }
    ];

    testActions = store.getActions();

    return store.dispatch(actions.deleteTodo(deleteObj))
      .then(() => {
        expect(testActions).toEqual(expectedActions);
      });
  });

  it('should create a getTodosSuccess action', () => {
    const data = [{ text: '1', id: '1' }, { text: '2', id: '2' }];
    const expectedAction = {
      type: TYPES.GET_TODOS_SUCCESS,
      data
    };

    testActions = actions.getTodosSuccess(data);

    expect(testActions).toEqual(expectedAction);
  });
});
