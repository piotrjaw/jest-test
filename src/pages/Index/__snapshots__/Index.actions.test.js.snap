// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index actions should create a getTodosSuccess action 1`] = `undefined`;

exports[`Index actions should create an action to create a todo 1`] = `undefined`;

exports[`Index actions should create an action to delete a todo 1`] = `undefined`;

exports[`Index actions should create an action to get all the todos 1`] = `undefined`;

exports[`Index actions should create an action to udpate a todo 1`] = `undefined`;

exports[`Index actions should reject on lack of text 1`] = `undefined`;
